mamba_ssm/ops/selective_scan_interface_ compilable.py

```diff
@@ -0,0 +1,272 @@
+import torch
+import torch.nn.functional as F
+from einops import rearrange
+from typing import Optional, Tuple
+
+import selective_scan_cuda
+
+
+@torch.library.custom_op(
+    "custom_ops::selective_scan_fwd",
+    device_types=["cuda"],
+    mutates_args=(),
+)
+def custom_selective_scan_fwd(
+    u: torch.Tensor,
+    delta: torch.Tensor,
+    A: torch.Tensor,
+    B: torch.Tensor,
+    C: torch.Tensor,
+    D: Optional[torch.Tensor],
+    z: Optional[torch.Tensor],
+    delta_bias: Optional[torch.Tensor],
+    delta_softplus: bool,
+    return_last_state: bool,
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor, bool, bool, bool]:
+    pass
+
+@torch.library.register_fake("custom_ops::selective_scan_fwd")
+def custom_selective_scan_fwd_fake(
+    u,
+    delta,
+    A,
+    B,
+    C,
+    D,
+    z,
+    delta_bias,
+    delta_softplus,
+    return_last_state,
+):
+    final_out = torch.empty_like(u)
+    dstate = A.size(1) * (2 if A.is_complex() else 1)
+    last_state_fake = u.new_empty((u.size(0), u.size(1), dstate)) if return_last_state else u.new_empty(0)
+    out_fake = torch.empty_like(u)
+    x_fake = u.new_empty((u.size(0), u.size(1), u.size(2), 2 * dstate))
+    return final_out, last_state_fake, out_fake, x_fake, False, False, z is not None
+
+@torch.library.register_kernel("custom_ops::selective_scan_fwd", "cuda")
+def custom_selective_scan_fwd_cuda(
+    u: torch.Tensor,
+    delta: torch.Tensor,
+    A: torch.Tensor,
+    B: torch.Tensor,
+    C: torch.Tensor,
+    D: Optional[torch.Tensor],
+    z: Optional[torch.Tensor],
+    delta_bias: Optional[torch.Tensor],
+    delta_softplus: bool,
+    return_last_state: bool,
+):
+    if u.stride(-1) != 1:
+        u = u.contiguous()
+    if delta.stride(-1) != 1:
+        delta = delta.contiguous()
+    if D is not None:
+        D = D.contiguous()
+    if B.stride(-1) != 1:
+        B = B.contiguous()
+    if C.stride(-1) != 1:
+        C = C.contiguous()
+    if z is not None and z.stride(-1) != 1:
+        z = z.contiguous()
+
+    squeeze_B = False
+    if B.dim() == 3:
+        B = rearrange(B, "b dstate l -> b 1 dstate l").contiguous()
+        squeeze_B = True
+
+    squeeze_C = False
+    if C.dim() == 3:
+        C = rearrange(C, "b dstate l -> b 1 dstate l").contiguous()
+        squeeze_C = True
+
+    out, x, *rest = selective_scan_cuda.fwd(u, delta, A, B, C, D, z, delta_bias, delta_softplus)
+    has_z = z is not None
+    final_out = rest[0].clone() if has_z else out.clone()
```

Comment:
Why are you cloning the tensor right here?

Reply:
Without the extra clone we get  (not in the test but on a real training session)
```python
RuntimeError: selective_scan_fwd (with implementation in <module 'torch._library.custom_ops' from '/opt/conda/lib/python3.11/site-packages/torch/_library/custom_ops.py'>): The output of this custom operator (1) must not also be an input to this custom operator and (2) may not alias any inputs to this custom operator or other returns. The most common way to trigger this error is if we have y = custom_op(x) and y and x are the same Tensor. Please instead return a clone of the offending output tensor(s) (e.g. return x.clone()) or refactor the custom operator to not return y.
```

Reply:
Huh that seems weird to me no? In the CPP code we are clearly creating a new tensor for out and out_z which are independent from any input tensor.

Reply:
it is coming from here:
https://github.com/pytorch/pytorch/blob/main/torch/_library/utils.py#L349C5-L372

Reply:
> aliasing...  other returns.

So I think that one candidate is that the same `final_out` return is aliasing  different buffers right?